protocols/smtp/src/main/java/org/apache/james/protocols/smtp/hook/RcptHook.java

```diff
@@ -28,12 +29,27 @@
     
     /**
      * Return the HookResult after run the hook
+     *
+     * @deprecated Use {@link #doRcpt(SMTPSession, MaybeSender, MailAddress)} instead
      * 
      * @param session the SMTPSession
      * @param sender the sender MailAddress
      * @param rcpt the recipient MailAddress
      * @return HookResult
      */
-    HookResult doRcpt(SMTPSession session, MailAddress sender, MailAddress rcpt);
+    @Deprecated
+    default HookResult doRcpt(SMTPSession session, MailAddress sender, MailAddress rcpt) {
+        return doRcpt(session, MaybeSender.of(sender), rcpt);
+    }
+
+    /**
+     * Return the HookResult after run the hook
+     *
+     * this strongly typed version smoothly handled null sender and should be prefered.
+     */
+    @SuppressWarnings("deprecated")
+    default HookResult doRcpt(SMTPSession session, MaybeSender sender, MailAddress rcpt) {
+        return doRcpt(session, sender.asOptional().orElse(null), rcpt);
```

Comment:
I must miss something, but when I read this I see that `doRcpt(SMTPSession, MaybeSender, MailAddress)` is calling `doRcpt(SMTPSession, MailAddress, MailAddress)` which is calling... `doRcpt(SMTPSession, MaybeSender, MailAddress)`...
please tell me I'm wrong :)

Reply:
You are not.

This cross default interface is designed to not break API for implementers who only need to implement just one method: either the new OR the old.

The downside here is that you sacrify compile time check for forcing the implementer to implement one of those method and will have by default an infinite loop.

I already described this choice in the base PR